pkg/daemon/ceph/osd/daemon.go

```diff
@@ -256,15 +261,10 @@ func getDataDirs(context *clusterd.Context, kv *k8sutil.ConfigMapKVStore, desire
 	}
 
 	// the osd dirs map doesn't exist yet
-	if len(dirList) == 0 {
-		// no dirs have been specified
-		if devicesSpecified {
-			// user is using devices instead of dirs
-			return map[string]int{}, removedDirs, nil
-		}
 
-		// no devices or dirs specified, return the default data dir
-		return map[string]int{context.ConfigDir: unassignedOSDID}, removedDirs, nil
```

Comment:
Help me understand what changed here. It seems we were already returning the default dir if there were zero directories and devices specified.

Reply:
that's correct, we were returning the default dir if there were no devices and no dirs specified...in the case where there were no existing created OSDs, and that was the problem.  If there were existing OSD dirs, we'd go into the `getActiveAndRemovedDirs` function.  In the case of the "default dir" being used, we'd have:
* the default dir OSD in the saved dir map
* 0 user specified dirs
In that case, we would accidentally think the default dir OSD should be removed, because the user didn't say to use any dirs!

The right thing to do here is to always have a dir list of at least the default directory when no devices or dirs are specified.  If there's an existing default dir OSD then good, it won't be removed.  If there are no existing OSDs then good, one will get created in the default dir.

I think this is the right approach for all scenarios.  Let me know if that doesn't explain it all.